### PR TITLE
Port roslyn-analyer style rules to experimental rules in roslyn.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
These are behind options marked as 'Experimental' in the UI.  

image

This is virtually a straight port, except to move onto the roslyn primitives we use insetad of the ones that Roslyn-Analyzers uses.  I've marked points of interest.

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveBracePlacement/CSharpConsecutiveBracePlacementDiagnosticAnalyzer.cs
Comment (Youssef1313) [inline src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveBracePlacement/CSharpConsecutiveBracePlacementDiagnosticAnalyzer.cs:20]: I think EnforceOnBuildValues should be used instead.

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs
Comment (Youssef1313) [inline src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:161]: If these are expected to be non-experimental after some time, won't we have to use other IDs, which is breaking?

Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:161]: 1.  Anything experimental doesn't have to worry about breaking. That's the point of being experimental. 
2. I don't see why we'd need to change the ids...

Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:161]: > 2. I don't see why we'd need to change the ids...

My point on this was that if you'll keep the 2000 range for experimental, they ids will need to change to something else. But that should be okay based on:

> Anything experimental doesn't have to worry about breaking. That's the point of being experimental.

Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:161]: Yes. It should be ok. What I meant though was simply to put these rules in the 2000 range apart from everything else.  I can envision lots of these rules, and if like them separated from the rest.

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveBracePlacement/CSharpConsecutiveBracePlacementDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveBracePlacement/CSharpConsecutiveBracePlacementDiagnosticAnalyzer.cs]: done.

**Review by sharwell — CHANGES_REQUESTED** [review]
I'm not a fan of this change for one primary reason: it moves rules into dotnet/roslyn because our UI implementation does not support presentation and configuration of rules coming from third-party sources (i.e. NuGet packages outside of this repository). Before migrating these rules, I would like to see the gap between first-party and third-party analyzer providers closed for a consistent experience that does not actively penalize authors on the grounds that they are not Microsoft employees.

/cc @jmarolf and marking for design review.

**Comment by Youssef1313** [comment]
> it moves rules into dotnet/roslyn because our UI implementation does not support presentation and configuration of rules coming from third-party sources

If I understand correctly, this wasn't the reason for moving the rules here from dotnet/roslyn-analyzers to dotnet/roslyn. See https://github.com/dotnet/roslyn-analyzers/pull/4647#issuecomment-754297131

I'm certainly with closing the gap for UI implementation between first-party and third-party analyzers, but I think this shouldn't block this PR?

**Comment by sharwell** [comment]
@Youssef1313 the blocker here is just that we shouldn't merge it before the next design review meeting

File: src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs
Comment (Youssef1313) [inline src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs:38]: This option feels unnecessary. If the user wants to allow multiple blank lines, why not simply disable this rule?

File: src/Analyzers/VisualBasic/Tests/NewLines/ConsecutiveStatementPlacement/ConsecutiveStatementPlacementTests.vb
Comment (Youssef1313) [inline src/Analyzers/VisualBasic/Tests/NewLines/ConsecutiveStatementPlacement/ConsecutiveStatementPlacementTests.vb:223]: Should the following code have the same behavior?

```vb
        Do While True
        Loop
```

File: src/Analyzers/VisualBasic/Analyzers/NewLines/ConsecutiveStatementPlacement/VisualBasicConsecutiveStatementPlacementDiagnosticAnalyzer.vb
Comment (Youssef1313) [inline src/Analyzers/VisualBasic/Analyzers/NewLines/ConsecutiveStatementPlacement/VisualBasicConsecutiveStatementPlacementDiagnosticAnalyzer.vb:41]: The case in https://github.com/dotnet/roslyn/pull/50358/files#r554634136 might fail due to missing check for `LoopStatementSyntax`.

```suggestion
            Return TypeOf node Is EndBlockStatementSyntax OrElse TypeOf node Is NextStatementSyntax OrElse TypeOf node Is LoopStatementSyntax
```

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs
Comment (Youssef1313) [inline src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs:41]: nit:

```suggestion
            => node is BlockSyntax or SwitchStatementSyntax;
```

**Comment by mavasani** [comment]
> it moves rules into dotnet/roslyn because our UI implementation does not support presentation and configuration of rules coming from third-party sources 

@sharwell This is not true. The formatting rules being added here never belonged to roslyn-analyzers repo. Rule added in analyzers repo were always meant to be a temporary solution to enable quick dogfooding. These rules ought to be deleted from roslyn-analyzers repo and rightfully belong to this repo as the primary source for formatting rules shipping with VS/.NET SDK. I am fine if you prefer to block this PR if you prefer an alternate implementation, but the rules will soon be removed from Roslyn.Diagnostics.Analyzers NuGet package in roslyn-analyzers repo.

File: src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs:38]: I don't understand your question. That's exactly what this check is for. If they disable the rule then this is what makes us bail out

**Comment by sharwell** [comment]
> but the rules will soon be removed from Roslyn.Diagnostics.Analyzers NuGet package in roslyn-analyzers repo

Roslyn.Diagnostics.Analyzers is the package for objective code style rules implemented for use in dotnet/roslyn without needing to go through a design review for public consumption in other projects. It's fine to relocate rules from that to another location, provided another location meets those same goals regarding minimal barrier to entry and suitable publication for consumption in dotnet/roslyn builds.

**Comment by mavasani** [comment]
> Roslyn.Diagnostics.Analyzers is the package for objective code style rules implemented for use in dotnet/roslyn without needing to go through a design review for public consumption in other projects

@sharwell This is not correct. This package is meant for rules that are very specific to Roslyn.sln, it is not a package to put unrelated rules, especially style rules which have nothing specific to Roslyn.sln. It has been unanimously agreed in prior design meeting that rules that the formatting/style rules added by Cyrus are extremely valuable to all customers, and need to be moved to Roslyn repo so they can ship out of the box and can be used by external customers. Currently, all enhancements to formatter or new formatting rules have been blocked for at least past 4-5 years on the basis of a new extensible formatting engine, which seems very unlikely to happen anytine soon. I like the approach taken by @CyrusNajmabadi in this PR to introduce these as experimental rules in CodeStyle package, and (if/when) in future we implement the proposed formatting engine extension, they can be ported over to these rules.

**Comment by CyrusNajmabadi** [comment]
> provided another location meets those same goals regarding minimal barrier to entry and suitable publication for consumption in dotnet/roslyn builds

I think that holds here. These are totally normal analyzer/fixers.  Anyone could write these themselves and publish. The idea here is simply that we think:

1. It would be nice to be on roslyn
2. We don't 100% want to commit to the rule and the current behavior being checked in. We'd like to be able to tweak things for a while. 

Once we're happy that we got things right, we can graduate this to a full fledged feature. 

--

Note: we do have ui for this. But it's completely optional. We could just add the code and simply document online and do nothing in the ui.  However, ui support was trivial and consistent with the rest of roslyn, so I did it. :-)

File: src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs
Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs]: I meant that the user can do that `using dotnet_diagnostic.IDExxxx.severity = none`. So why an option that does the same thing needed?

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs:41]: thanks!

File: src/Analyzers/VisualBasic/Analyzers/NewLines/ConsecutiveStatementPlacement/VisualBasicConsecutiveStatementPlacementDiagnosticAnalyzer.vb
Comment (CyrusNajmabadi) [reply src/Analyzers/VisualBasic/Analyzers/NewLines/ConsecutiveStatementPlacement/VisualBasicConsecutiveStatementPlacementDiagnosticAnalyzer.vb:41]: yup.  good catch!

File: src/Analyzers/VisualBasic/Tests/NewLines/ConsecutiveStatementPlacement/ConsecutiveStatementPlacementTests.vb
Comment (CyrusNajmabadi) [reply src/Analyzers/VisualBasic/Tests/NewLines/ConsecutiveStatementPlacement/ConsecutiveStatementPlacementTests.vb:223]: yup!

File: src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/NewLines/MultipleBlankLines/AbstractMultipleBlankLinesDiagnosticAnalyzer.cs]: I guess i'm not 100% sure what checks still need to be in the analyzer.  It seems if the user explicitly sets the value to `true:...` we would need to still see that and bail out, no?  @mavasani for clarity.

**Comment by JoeRobich** [comment]
Notes from Design Review:
- Propose that the rules being added would follow the naming style used by our current .editorconfig style options
- Propose vetting these rules with BCL team that they do not go against guidance
- Investigate implementing style rules as independent formatting rules

**Comment by CyrusNajmabadi** [comment]
@sharwell so can we merge this in, then port to formatting rules when possible?

**Comment by CyrusNajmabadi** [comment]
@sharwell  I looked into making these into formatting rules and i immediately ran into snags.  Specifically i was trying to do the formatting rule to affect formatting in cases liek this:

```c#
    }

}
```

Following hte pattern around GetAdjustNewLinesOperationBetweenMembers, i created a rule to force the blank lines between consecutive braces to 1-max.  

However, while the rules triggered properly and added htese constraints, later portions of the formatting engine then undid this work.  Specifically, they would suppress the rule because they woudl see things like this:

```c#
    } // foo

}
```

And they would say: "ah, i can't touch the formatting here because there's a comment".

This is the sort of thing that makes it extremely challenging to put this type of operation into the formatting engine.  Because the system works through a combination of producing constraints and then deciding what to do with the constraints, it's extremely difficult to figure out how to even get rules like this working. Should there be a new type of constraint that works a different way?  Should the constraint interpreter be changed to understand these scenarios?  If we do that, will that affect other rules, etc. etc.?

I'm strongly supportive of a good extensible formatting system to help enable these scenarios.  However, until we get that, I would like to get these experimental rules in *now* so we can at least benefit from them for our engineering systems.

**Comment by sharwell** [comment]
> And they would say: "ah, i can't touch the formatting here because there's a comment".

Yes, this is by design. There are regions of code that the formatter cannot touch (collection initializers, suppressed-format regions, and cases like the comment above). The rules here must adhere to these policies.

**Comment by CyrusNajmabadi** [comment]
> but for now those would not block conversion to formatting rules.

But it literally did block the conversion. Because I could not connect this to a formatting rule. 

Look. I get what you want here. But please let's just move forward and get something in place now that we can use to enforce the desired style until you're able to wrangle the formatting engine into a better place. 

I don't want this to just not go anywhere for months or years because we're not willing to take a pragmatic step now while we can. 

@mavasani on this as well.

**Comment by CyrusNajmabadi** [comment]
> Yes, this is by design. 

Then the formatting engine is not in a suitable place to take on these rules, and we should not block this pr asking it to accomplish its task by sitting in the formatting engine. 

Please unblock this so it can be merged in. When we have a more suitable formatting engine, we can port these to be rules with full fidelity.

**Comment by mavasani** [comment]
I fully agree with @CyrusNajmabadi here and appreciate the attempt to implement this in the formatter. I think we should go 2 stage route here - stage 1 add these as special formatting analyzer experimental rules (either all of them with a single ID or separate ID per rule, both seem fine to me) and stage 2 ports these rules to formatter engine.

**Comment by sharwell** [comment]
I will remove my block with the comment that I do not believe these rules have been implemented in a manner that meets the fit-and-finish expectations I would have for a formatting feature in the IDE context, and I do not agree that the ship priority for this feature outweighs expectations here.

**Comment by mavasani** [comment]
> I do not agree that the ship priority for this feature outweighs expectations here.

@sharwell Sorry, but I strongly disagree. This feature is much more valuable than the desire to make it fit the Ctrl+K+D fit and finish experience versus Ctrl+Dot+Enter experience.

The core issues here are that the expectations have been defined by an engineer, with no input from PM or end users whether they would expect all whitespace/newline related rules to be under a single umbrella OR would be fine with a hybrid approach where newer (likely breaking) additions are experimental analyzers that helps them get new features to improve their code base with build enforcement versus delay it because the fit and finish of mechanisms of invoking the fix are not in line with existing whitespace formatting experience. For example, we hit these cases multiple times in PRs to roslyn-analyzers, where such a feature coming out of VS/.NET SDK would be extremely valuable. Also note that once the rule is enabled on build and the repo is clean, as an end user, I really don't care if I need to do a Ctrl+K+D or a Ctrl+Dot+Enter to fix new warnings while typing - both are expected as well known modes of fixing diagnostics in the IDE.

Additionally, this block of adding such newer, valuable rules has been present long enough (at least 5+ years now), with very little to no progress on the formatter engine or formatting API side. Unless you can unblock Cyrus in his attempts to follow your suggestions, I do not see a clear path forward here.

**Comment by mavasani** [comment]
I am also wondering if we can make analyzers themselves as the extension point for formatting engine instead of a formatting API.

1. Allow analyzers to be annotated with a special attribute, say `FormattingAnalyzerAttribute`
2. Formatting engine runs these special formatting analyzers and their fix at the same point where we were expecting to run the formatting engine extensions in the design where we have new formatting API/extensibility.
3. Design how to deal with conflicts and ordering of these formatting analyzers/fixes the same way we would do for formatting engine API/extensions.

Another big advantage here would be that Roslyn developers are already familiar with writing analyzers, as opposed to learning a whole new API and system for writing formatting rules/extensions that directly plug into the formatting engine.

I am not sure about the overall feasibility of this approach as I don't know the formatting engine well enough though.

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs
Comment (mavasani) [inline src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs:41]: All of these would ideally be on `ISyntaxFacts` and can be used from the base type itself.

**Comment by CyrusNajmabadi** [comment]
> I will remove my block with the comment that I do not believe these rules have been implemented in a manner that meets the fit-and-finish expectations I would have for a formatting feature in the IDE context

That's why tehse are explicitly experimental.  It allows us to unblock and get something valuable out for the team (which we've already been doing), without burden on roslyn-analyzers.

I would absolutely be happy to have this come in through the formatting pipeline.  However, i see no effective way to do that in either the short or medium term, and blocking this useful *experimental* feature because of that isn't beneficial to our team.

File: src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/NewLines/ConsecutiveStatementPlacement/CSharpConsecutiveStatementPlacementDiagnosticAnalyzer.cs:41]: indeed.  i will fixup

**Comment by CyrusNajmabadi** [comment]
> I am also wondering if we can make analyzers themselves as the extension point for formatting engine instead of a formatting API.

**Comment by CyrusNajmabadi** [comment]
> I am also wondering if we can make analyzers themselves as the extension point for formatting engine instead of a formatting API.

I was thinking about this.  I also gave some thought to the idea that the formatting engine could have a postpass where you just did tree transformations, and i'd put tehse rules in that bucket.  THe problem that arose was not about ordering (though that is a bit of a concern), but is about the currency the formatting engine deals with.  Specifically, the formatting engine doesn't like to produce the final 'tree/root' unless asked.  Instead, it produces edits to text.  So it's a challenge to incorporate a general mechanism to take those edits and combine them with tree transforms.

That said, it's perhaps possible we could merge those edits with the edits we 'diff' out from the changes the formatting analyzer made.  If they didn't 'collide' or anything, we migth be able to make a combined set.

**Comment by jmarolf** [comment]
> > I am also wondering if we can make analyzers themselves as the extension point for formatting engine instead of a formatting API.
> 
> I was thinking about this. I also gave some thought to the idea that the formatting engine could have a postpass where you just did tree transformations, and i'd put tehse rules in that bucket. THe problem that arose was not about ordering (though that is a bit of a concern), but is about the currency the formatting engine deals with. Specifically, the formatting engine doesn't like to produce the final 'tree/root' unless asked. Instead, it produces edits to text. So it's a challenge to incorporate a general mechanism to take those edits and combine them with tree transforms.
> 
> That said, it's perhaps possible we could merge those edits with the edits we 'diff' out from the changes the formatting analyzer made. If they didn't 'collide' or anything, we migth be able to make a combined set.

I also have thoughts forming in my mind about this. A lot of my concerns are about tying it to Ctlr+K+D. If custom rules could be tied to some other command (such as "code cleanup"). I think an analyzer API approach would be easiest for customers to write against (since they could just re-use the style cop analyzers instead of needing to re-implement them). This also makes it easier for things like `dotnet format` to work since today we just run the fix-all for  stylecop.

**Comment by CyrusNajmabadi** [comment]
@mavasani are yo uok with thsi going in?  It's just ports of the same work in roslyn-analyzers.  Thanks!

**Review by mavasani — APPROVED** [review]
Sorry for the delay in review, yes this looks good to port to me.

**Comment by mavasani** [comment]
@CyrusNajmabadi Would you be able to create a PR or file an issue in roslyn-analyzers repo to remove the ported analyzers from there? Thanks.

**Comment by CyrusNajmabadi** [comment]
Yup.  I'll remove once this goes in.

**Pull request merged**
